File: src/kuka_position_control/scripts/circle_motion.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
import math
import time

class CircularMotionPublisher(Node):
    def __init__(self):
        super().__init__('circular_motion_publisher')
        self.publisher_ = self.create_publisher(TransformStamped, '/target_pose', 10)
        self.timer_period = 0.1  # seconds
        self.i = 0
        self.t_0 = self.get_clock().now().to_msg().nanosec * 1e-9
        self.timer = self.create_timer(self.timer_period, self.timer_callback)

    def timer_callback(self):
        msg = TransformStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        t = time.time() - self.t_0
        msg.transform.translation.x = math.cos(t)
        msg.transform.translation.y = math.sin(t)
        msg.transform.translation.z = 0.0
        msg.transform.rotation.x = 0.0
        msg.transform.rotation.y = 0.0
        msg.transform.rotation.z = 0.0
        msg.transform.rotation.w = 1.0
        self.publisher_.publish(msg)
        self.i += 1

def main(args=None):
    rclpy.init(args=args)
    circular_motion_publisher = CircularMotionPublisher()
    rclpy.spin(circular_motion_publisher)
    circular_motion_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/kuka_position_control/src/circle_motion.cpp
#include <ros/ros.h>
#include <geometry_msgs/TransformStamped.h>
#include <cmath>

int main(int argc, char **argv){
  ros::init(argc, argv, "circle_publisher");
  ros::NodeHandle nh;
  ros::Publisher pub = nh.advertise<geometry_msgs::TransformStamped>("/target_pose", 1000);

  ros::Rate rate(10); // 100 Hz
  double t = 0.0;
  double r = 0.3;
  geometry_msgs::TransformStamped center;
  center.header.stamp = ros::Time::now();
  center.header.frame_id = "world";
  center.child_frame_id = "target";
  center.transform.translation.x = 1.2; // Circular motion in X
  center.transform.translation.y = 0; // Circular motion in Y
  center.transform.translation.z = 0.5; // No motion in Z
  center.transform.rotation.x = 0;
  center.transform.rotation.y = 0.707;
  center.transform.rotation.z = 0;
  center.transform.rotation.w = 0.707; // No rotation

  pub.publish(center);

  for(int i = 0; i < 30; i++);
    rate.sleep();
  

  while(ros::ok() && t < 100){
    geometry_msgs::TransformStamped msg;
    msg.header.stamp = ros::Time::now();
    msg.header.frame_id = "world";
    msg.child_frame_id = "target";
    msg.transform.translation.x = 1.2 + r * sin(t * 2 * M_PI / 10.0); // Circular motion in X
    msg.transform.translation.y = r * cos(t * 2 * M_PI / 10.0); // Circular motion in Y
    msg.transform.translation.z = 0.5; // No motion in Z
    msg.transform.rotation.x = 0;
    msg.transform.rotation.y = 0.707;
    msg.transform.rotation.z = 0;
    msg.transform.rotation.w = 0.707; // No rotation

    pub.publish(msg);
    
    t += 0.1; // Increase the time step
   
    ros::spinOnce();
    rate.sleep(); // Sleep to maintain the desired rate
  }

  if(ros::ok())
  {
    pub.publish(center);
  }

  return 0;
}
